src/features/nameCleansing/nameCleansing.ts

```diff
@@ -0,0 +1,71 @@
+import { GuildMember } from 'discord.js'
+
+import { clean } from 'unzalgo'
+
+import { Environment } from '@/config'
+import { BotContext } from '@/types/BotContext'
+
+import {
+  checkDehoisted,
+  checkExclamationMark,
+  checkNameAgainstPatterns,
+  checkZalgo,
+} from './checkName'
+
+export const compiled = new Map<string, RegExp>()
+
+export async function checkName(member: GuildMember, botContext: BotContext) {
+  const { log, runtimeConfiguration } = botContext
+  const config = runtimeConfiguration.data.nameCleansing
+  const configNameList = runtimeConfiguration.data.nameChecker
+  if (member.guild.id !== Environment.GUILD_ID) return
+
+  const {
+    enabled,
+    enabledCheckZalgo,
+    enabledCheckDehoisted,
+    enabledCheckExclamationMark,
+    enabledCheckBadName,
+  } = config
+  if (!enabled) {
+    return
+  }
+
+  const { patterns } = configNameList
+  if (!patterns) {
+    return
+  }
+
+  if (checkZalgo(member.displayName) && enabledCheckZalgo) {
+    member
+      .setNickname(clean(member.displayName), 'Zalgo Name Detacted')
+      .catch(() => {
+        log.info(`Can't change ${member.displayName}'s name`)
+      })
+  }
+
+  if (checkDehoisted(member.displayName) && enabledCheckDehoisted) {
+    member
+      .setNickname(member.user.username, 'Dehoisted Name Detacted')
+      .catch(() => {
+        log.info(`Can't change ${member.displayName}'s name`)
+      })
+  }
+
+  if (checkExclamationMark(member.displayName) && enabledCheckExclamationMark) {
+    member
+      .setNickname(member.user.username, 'Exclamation Mark Name Detacted')
+      .catch(() => {
+        log.info(`Can't change ${member.displayName}'s name`)
+      })
+  }
+
+  if (
+    checkNameAgainstPatterns(member.displayName, patterns, log) &&
+    enabledCheckBadName
+  ) {
+    member.setNickname(member.user.username, 'Bad Name Detacted').catch(() => {
+      log.info(`Can't change ${member.displayName}'s name`)
+    })
+  }
```

Comment:
This part of code checks for certain name pattern and member the name to their user name. However, if the member’s default username also matches the rules here, the bot may run into an infinite loop krub.

Suggested change:

- In these cases, please check against `nickname` instead of `displayName`. 

  ![image](https://github.com/kaogeek/kaogeek-discord-bot/assets/193136/96c1e6bb-c882-4f4a-a20f-df84535db808)

- Since the code here only works if user has a `nickname`, we can wrap the code in here with `if (member.nickname != null)`

- To reset nickname, use `null` instead of `member.user.username`.

  ![image](https://github.com/kaogeek/kaogeek-discord-bot/assets/193136/6128b44f-a6cb-4f4f-9f83-e5dd70b6b0c3)

Reply:
Done krub